HEN_HOUSE/egs++/geometry/egs_autoenvelope/volcor.h

```diff
@@ -0,0 +1,663 @@
+    void outputResults() const {
+
+        egsInformation(" --------- Volume Correction Results -----------\n");
+        egsInformation(" Time taken                  = %.4f s (%.3E s/point) \n", time, time/npoints);
+        egsInformation(" Density of points used      = %.3E points/cm^-3\n", density);
+        egsInformation(" Number of points used       = %G\n", npoints);
+        egsInformation(" Bounding shape volume       = %.5E cm^3\n", bounds_volume);
+        egsInformation(" Inscribed geometry volume   = %.5E cm^3\n", inscribed_volume);
+        options->rng->describeRNG();
+        egsInformation(" -----------------------------------------------\n");
+
+    }
+
+};
+
+
+/*! \brief Apply volume corrections to base regions.
+ *
+ * After the MC volume run is complete, use the tallied hits and mode requested
+ * to set the correct volume of the base regions. */
+vector<EGS_Float> applyVolumeCorrections(VCOptions *opts, HitCounterT hit_counter, vector<EGS_Float> uncorrected) {
+
+    bool zero = opts->mode == ZERO_VOL;
+    double bounds_vol = opts->bounds_volume;
+    double npoints =  opts->npoints;
+    vector<EGS_Float> corrected_vols(uncorrected);
+
+    for (HitCounterT::iterator hi = hit_counter.begin(); hi != hit_counter.end(); hi++) {
+
+        int base_reg = hi->first;
+        if (base_reg < 0) {
+            continue;
+        }
+
+        int hits = hi->second;
+        EGS_Float corrected = uncorrected[base_reg] - bounds_vol*double(hits)/npoints;
+        corrected_vols[base_reg] = zero ? 0 : corrected;
+    }
+
+    return corrected_vols;
+}
+
+/*! \brief Apply file volume corrections to base regions.
+ *
+ * After volumes are loaded from file use the volumes and mode requested
+ * to set the correct volume of the base regions. */
+vector<EGS_Float> applyFileVolumeCorrections(VCOptions *opts, vector<RegVolume> &reg_volumes, vector<EGS_Float> uncorrected) {
+
+    bool zero = opts->mode == ZERO_VOL;
+    vector<EGS_Float> corrected_vols(uncorrected);
+
+    for (size_t rvc=0; rvc < reg_volumes.size(); rvc++) {
+        int base_reg = reg_volumes[rvc].first;
+        EGS_Float vol =  reg_volumes[rvc].second;
+        corrected_vols[base_reg] = zero ? 0 : vol;
+    }
+
+    return corrected_vols;
+}
+
+vector<EGS_Float> getUncorrectedVolumes(EGS_BaseGeometry *base) {
+    vector<EGS_Float> uncorrected;
+    for (int ir=0; ir < base->regions(); ir++) {
+        EGS_Float vol = base->getMass(ir)/base->getRelativeRho(ir);
+        uncorrected.push_back(vol);
+    }
+    return uncorrected;
+
+}
+
+/*! \brief Run the MC simulation.
+ *
+ * Finds regions with inscribed geometries and corrects
+ * the volume of those regions. The algorithm is described
+ * on main page of the Auto Envelope documentation. */
+VCResults findRegionsWithInscribed(VCOptions *opts, EGS_BaseGeometry *base,
+                                   vector<EGS_BaseGeometry *> inscribed, vector<EGS_AffineTransform *> transforms) {
```

Comment:
We should look into making use of the more efficient volume calculations in pull request #184.